src/trans-word-level/instantiate_word_level.cpp

```diff
@@ -175,24 +169,131 @@ wl_instantiatet::instantiate_rec(exprt expr, const mp_integer &t) const
 
       // Do we exceed the bound? Make it 'true'
       if(upper >= no_timeframes)
-        return {no_timeframes - 1, true_exprt()};
+        return {{no_timeframes - 1, true_exprt()}};
 
-      exprt::operandst disjuncts;
+      std::vector<std::pair<mp_integer, exprt>> match_points;
 
       for(mp_integer u = lower; u <= upper; ++u)
       {
-        disjuncts.push_back(
-          instantiate_rec(sva_cycle_delay_expr.op(), u).second);
+        auto sub_result = instantiate_sequence(sva_cycle_delay_expr.op(), u);
+        for(auto &match_point : sub_result)
+          match_points.push_back(match_point);
       }
 
-      return {upper, disjunction(disjuncts)};
+      return match_points;
     }
   }
-  else if(expr.id()==ID_sva_sequence_concatenation)
+  else if(
+    expr.id() == ID_sva_sequence_concatenation ||
+    expr.id() == ID_sva_overlapped_implication ||
+    expr.id() == ID_sva_non_overlapped_implication)
+  {
+    auto &implication = to_binary_expr(expr);
+    std::vector<std::pair<mp_integer, exprt>> result;
+
+    // This is the product of the match points on the LHS and RHS
+    const auto lhs_match_points = instantiate_sequence(implication.lhs(), t);
+    for(auto &lhs_match_point : lhs_match_points)
+    {
+      // The RHS of the non-overlapped implication starts one timeframe later
+      auto t_rhs = expr.id() == ID_sva_non_overlapped_implication
+                     ? lhs_match_point.first + 1
+                     : lhs_match_point.first;
+
+      // Do we exceed the bound? Make it 'true'
+      if(t_rhs >= no_timeframes)
+        return {{no_timeframes - 1, true_exprt()}};
```

Comment:
Could `no_timeframes` be zero? (And: does it matter if it is, i.e., is it a problem if the result includes a negative number?)

Reply:
It's `bound + 1`, where `bound` is unsigned, so shouldn't. But could add a DATA_INVARIANT.